File: core/src/net/brenig/pixelescape/lib/CycleIntArray.java
package net.brenig.pixelescape.lib;

/**
 * int array that is arranged in a Ring<br></br>
 * when new elements get added the oldest elements get lost
 */
public class CycleIntArray {

	private int[] data;
	private int index;


	public CycleIntArray(int size, int initValue) {
		if(size < 1) {
			throw new IllegalArgumentException("The specified size has to be 1 or greater!");
		}
		data = new int[size];

		fill(initValue);
		index = size - 1;
	}

	private int convertToLocalIndex(int globalIndex) {
		return (index + globalIndex) % data.length;
	}

	private void updateIndexBounds() {
		index = index % data.length;
	}

	/**
	 * returns the object at the given index
	 * an index of 0 returns the oldest object, an index of size - 1 the newest
	 */
	public int get(int index) {
		return data[convertToLocalIndex(index)];
	}

	/**
	 * @see #get(int)
	 * @return the object at the given index reversed (--> index of 0 returns the newest object, an index of size-1 the oldest)
	 */
	public int getFromNewest(int index) {
		return get(data.length - index - 1);
	}

	public void add(int element) {
		index++;
		updateIndexBounds();
		data[index] = element;
	}

	public void set(int index, int element) {
		data[convertToLocalIndex(index)] = element;
	}

	public int getNewest() {
		return data[index];
	}

	public int getOldest() {
		return data[(index + 1) % data.length];
	}

	@Override
	public String toString() {
		int iMax = data.length - 1;
		StringBuilder b = new StringBuilder();
		b.append('[');
		for (int i = 0; ; i++) {
			b.append(String.valueOf(data[convertToLocalIndex(i)]));
			if (i == iMax)
				return b.append(']').toString();
			b.append(", ");
		}
	}

	public void fill(int value) {
		for(int i = 0; i < data.length; i++) {
			data[i] = value;
		}
	}
}

File: core/src/net/brenig/pixelescape/lib/CycleIntArray.kt
package net.brenig.pixelescape.lib

/**
 * int array that is arranged in a Ring<br></br>
 * when new elements get added the oldest elements get lost
 */
class CycleIntArray(size: Int, initValue: Int) {

    private val data: IntArray
    private var index: Int = 0

    val newest: Int
        get() = data[index]

    val oldest: Int
        get() = data[(index + 1) % data.size]


    init {
        if (size < 1) {
            throw IllegalArgumentException("The specified size has to be 1 or greater!")
        }
        data = IntArray(size)

        fill(initValue)
        index = size - 1
    }

    private fun convertToLocalIndex(globalIndex: Int): Int {
        return (index + globalIndex) % data.size
    }

    private fun updateIndexBounds() {
        index = index % data.size
    }

    /**
     * returns the object at the given index
     * an index of 0 returns the oldest object, an index of size - 1 the newest
     */
    operator fun get(index: Int): Int {
        return data[convertToLocalIndex(index)]
    }

    /**
     * @return the object at the given index reversed (--> index of 0 returns the newest object, an index of size-1 the oldest)
     * @see .get
     */
    fun getFromNewest(index: Int): Int {
        return get(data.size - index - 1)
    }

    fun add(element: Int) {
        index++
        updateIndexBounds()
        data[index] = element
    }

    operator fun set(index: Int, element: Int) {
        data[convertToLocalIndex(index)] = element
    }

    override fun toString(): String {
        val iMax = data.size - 1
        val b = StringBuilder()
        b.append('[')
        var i = 0
        while (true) {
            b.append(data[convertToLocalIndex(i)].toString())
            if (i == iMax)
                return b.append(']').toString()
            b.append(", ")
            i++
        }
    }

    fun fill(value: Int) {
        for (i in data.indices) {
            data[i] = value
        }
    }
}
